File: codigos/flash_minGibbs.m
function flash_minGibbs

    %% condicao do sistema 
    
    % temperatura
    T = 298.15;
    
    % fracao molar global dos componentes
    z = [.4,.6];
    
    % numero de mols total
    n = 1;
    
    % numero de mols dos componentes
    ni = z* n; 
       
    % numero de componentes
    nc=length(z);
    
    % numero de fases
    nf = 2;
        
    %% parametros UNIQUAC 

    %  1-butanol ~ agua
    r = [3.4543     .92 ];
    q = [3.0520  	1.4 ];
    q1 = q;
    aij = [ 0           -82.688;
        443.56      0 ;];
    tau = exp(-aij/T);
    
    %% encontrando o minimo global da energia de Gibbs com o metodo do enxame de particulas
    
    % parametros da funcao particleswarm                    
    nvars = 2;
    lim_inf = [1e-50, 1e-50];
    lim_sup = [ni(1)-1e-5, ni(2)-1e-5];
    
    % minimizando!
    [xmin,fmin,eflag,output] = particleswarm (@deltaG, nvars, lim_inf, lim_sup)
   
    %% plotando a energia de gibbs e seu otimo global
    
    % tamanho do passo
    step = .05;
        
    % malha
    n_grid1 = 1e-16: step : ni(1)-step;
    n_grid1 = [n_grid1 ni(1)-1e-5];
    n_grid2 = 1e-16: step : ni(2)-step;
    n_grid2 = [n_grid2 ni(2)-1e-5];
    
   % matriz que armazenara os valores de deltaG
    Z = zeros(length(n_grid1),length(n_grid2));
    
    % calculando o deltaG para cada ponto da malha
    for i1=1:length(n_grid1)
        for i2 = 1:length(n_grid2)
            Z(i1,i2) = deltaG([n_grid1(i1);n_grid2(i2)]);       
        end
    end
    
    % finalmente, plotando!
    
    figure(2)
    
    surf(n_grid2, n_grid1, Z);
    
    xlabel('mols de agua')
    ylabel('mols de 1-butanol')
    zlabel('\Delta G')
    title('Fase aquosa')
    
    % marcando o otimo
    hold on;
    plot3(xmin(2),xmin(1),fmin,'*r');   

    %% funcao objetivo para calculo de equilibrio: energia de gibbs
    
    function deltaG = deltaG(nij_vetor)
        
        nij = zeros(nc,nf-1);
        xij = zeros(nc,nf-1);
        ni_nf = zeros(nc,1);
        gamma = zeros(nc,nf-1);
        
        for i=1:nc
            for j=1:nf-1
                nij(i,j) = nij_vetor(j+(i-1)*(nf-1));
            end
        end
        
        deltaG = 0;
                
        for j = 1:nf-1
            for i = 1:nc               
                xij(i,j) = nij(i,j)/sum(nij(:,j));
            end
            gamma(:,j) = uniquac(xij(:,j)', r, q, q1, tau);
        end
                
        for j=1:nf-1
            for i = 1:nc
                deltaG = deltaG + nij(i,j)*log(xij(i,j)*gamma(i,j))/n;
            end
        end
        
        for i=1:nc
                ni_nf(i) = ni(i) - sum(nij(i,:));
        end
        
        xi_nf = ni_nf/sum(ni_nf);
        gamma_nf = uniquac(xi_nf', r, q, q1, tau);
                
        for i = 1:nc
            deltaG = deltaG + ni_nf(i)*log(xi_nf(i)*gamma_nf(i))/n;
        end
        
    end 
    
end

File: codigos/python/flash_minGibbs.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Jun  6 01:47:29 2017

@author: afranio
"""

import numpy as np
import scipy.optimize as opt
from thermo.activity import UNIQUAC
import matplotlib.pyplot as plt
from matplotlib import cm
from mpl_toolkits.mplot3d import Axes3D
#%matplotlib inline

# condicoes
TK = 298.15                 # temperatura
z = np.array([0.4, 0.6])    # composicao da alimentacao
nc = z.size                 # numero de componentes
nf = 2                      # numero de fases
n = 1                       # numero de mols total
ni = z*n                    # numeros de mols por componente

# parametros UNIQUAC
r = np.array([3.4543, 0.92])
q = np.array([3.0520, 1.4])
aij = np.array([[0,-82.688],[443.56,0]])
tau = np.exp(-aij/TK)

# preparando arrays que serao usadas na funcao objetivo
nij = np.zeros((nc,nf))
xij = np.zeros((nc,nf-1))
ni_nf = np.zeros(nc)
gamma = np.zeros_like(xij)

# funcao objetivo: energia de gibbs de mistura
def deltaG(nij_vetor):
    
    for j in range(0,nf-1):
        for i in range(0,nc):
            nij[i,j] = nij_vetor[j+i*(nf-1)]
    
    dG = 0
    
    for j in range (0,nf-1):
        for i in range (0,nc):
            xij[i,j] = nij[i,j]/sum(nij[:,j])
        gamma[:,j] = UNIQUAC(xij[:,j],r,q,tau)
        
    for j in range(0,nf-1):
        for i in range(0,nc):
            dG += nij[i,j]*np.log(xij[i,j]*gamma[i,j])/n
                     
    for i in range(0,nc):
        ni_nf[i] = ni[i] - sum(nij[i,:])
        
    xi_nf = ni_nf/sum(ni_nf)
    gamma_nf = UNIQUAC(xi_nf,r,q,tau)
    
    for i in range(0,nc):
        dG += ni_nf[i]*np.log(xi_nf[i]*gamma_nf[i])/n
    
    return dG

# encontrando o estado de equilibrio: minimo global de deltaG
bounds_phase1 = ((1e-50, ni[0]-1e-5), (1e-50, ni[1]-1e-5))
result = opt.differential_evolution(deltaG,bounds_phase1)

# ajustando resultados

dG_min = result.fun

ni_ph1 = result.x
ni_ph2 = ni - ni_ph1

xi_ph1 = ni_ph1/sum(ni_ph1)
xi_ph2 = ni_ph2/sum(ni_ph2)

B = sum(ni_ph1)/n
          
print('x[:,0]: ', xi_ph1[:])
print('x[:,1]: ', xi_ph2[:])
print('Beta:   ', B)
print('delta G:', dG_min)
          
# plotando deltaG e seu minimo global

#https://stackoverflow.com/questions/8722735/i-want-to-use-matplotlib-to-make-a-3d-plot-given-a-z-function

n1_axis = np.linspace(1e-16,ni[0]-1e-5,50)
n2_axis = np.linspace(1e-16,ni[1]-1e-5,50)
        
N1_axis, N2_axis = np.meshgrid(n1_axis, n2_axis)     

zs = np.array([deltaG([n1_axis, n2_axis]) for n1_axis,n2_axis in zip (np.ravel(N1_axis), np.ravel(N2_axis))])
Z = zs.reshape(N1_axis.shape)
        
fig = plt.figure()
ax = fig.gca(projection='3d')
ax.plot_surface(N1_axis, N2_axis, Z, cmap=cm.coolwarm)

ax.set_xlabel('Fase Aquosa')
ax.set_ylabel('Fase Orgânica')
ax.set_zlabel('$\Delta$ G')

# os dois minimos (o primeiro considerando a primeira fase como a organica, 
#o segundo considerando a primeira como aquosa)
ax.scatter(ni_ph1[0], ni_ph1[1], dG_min)
ax.scatter(ni_ph2[0], ni_ph2[1], dG_min)